### PR TITLE
Hot reload with OpenAPI support

### LINKED ISSUE #758: BentoML Artifacts Refactor: Support unload/load/pack dynamically
**Is your feature request related to a problem? Please describe.**

There have been a number of user requests related to loading and unloading models artifacts dynamically lately and this is to add this support.

Use cases:
* Speed up BentoService development workflow - currently user always needs to figure out how to save the model and load it back, then test it with the BentoService class. Ideally, the user should be able to do that with BentoML's model artifact class and build their BentoService class iteratively.
* Online learning - a new model is being training on the fly
* Composing model pipeline dynamically and loading models lazily 
* Automatically load new versions of artifact trained from a local directory, similar to the TensorFlow serving design https://github.com/bentoml/BentoML/issues/754 

**Describe the solution you'd like**

WIP

**Describe alternatives you've considered**
n/a

**Additional context**
n/a

### PR DESCRIPTION
**Pull request by akainth015**
## Description
Fixes #758 and adds hotswap support by switching to asynchronous operations for the server.

## Motivation and Context
Hotswapping trained models with BentoML is very useful for the reasons detailed in the issue above. This means they can be switched out even after the server has been started.

## How Has This Been Tested?
Still in progress

## Types of changes

- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Documentation
- [ ] Test, CI, or build
- [ ] None

## Components (if applicable)
- [ ] BentoService (model packaging, dependency management, handler definition)
- [ ] Model Artifact (model serialization, multi-framework support)
- [x] Model Server (mico-batching, logging, metrics, tracing, benchmark, OpenAPI)
- [x] YataiService (model management, deployment automation)
- [ ] Documentation

## Checklist:

- [ ] My code follows the bentoml code style, both `./dev/format.sh` and
  `./dev/lint.sh` script have passed
  ([instructions](https://github.com/bentoml/BentoML/blob/master/DEVELOPMENT.md#style-check-and-auto-formatting-your-code)).
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] My change requires a change in [bentoml/gallery](https://github.com/bentoml/gallery) example notebooks
- [ ] I have sent a pull request to [bentoml/gallery](https://github.com/bentoml/gallery) to make that change

### CODE REVIEW

**Comment by parano** [comment]
Hi @akainth015 - #758 is a work-in-progress proposal to introduce setup/loading/unloading hooks to the artifacts API so that individual artifacts can implement those methods and allow more customization. I will fill in more details on the design proposal and use cases on that issue soon. 

I'm not sure building an async version of BentoAPIServer solves that issue, but it is an interesting idea. BentoML currently has an async layer built with aiohttp that does request handling and micro-batching(https://docs.bentoml.org/en/latest/guides/micro_batching.html).  I think this can be useful when the user does not need micro-batching. Although the user may lose the ability to run multiple model workers.

File: bentoml/server/bento_api_server.py [hunk 1]
```diff
@@ -41,17 +40,20 @@
...
-<div id="swagger-ui-container"></div>
-<script src="/static/swagger-ui-bundle.js"></script>
-<script>
-    SwaggerUIBundle({{
-        url: '{url}',
-        dom_id: '#swagger-ui-container'
-    }})
-</script>
+    <div id="swagger-ui-container"></div>
+    <script src="https://unpkg.com/swagger-ui-dist@3/swagger-ui-bundle.js"></script>
```
Comment (parano) [inline bentoml/server/bento_api_server.py:50]: One benefit of not using CDN is that the user can use this web UI even without the internet. A few early users requested this a while back so we changed it including "swagger-ui-bundle.js" in /static and use it instead

Comment (akainth015) [reply bentoml/server/bento_api_server.py:50]: Yeah, I was having syntax errors with the local version of the script, not intended to be a permanent change (that's why I didn't change the style sheet)

**Comment by akainth015** [comment]
> Hi @akainth015 - #758 is a work-in-progress proposal to introduce setup/loading/unloading hooks to the artifacts API so that individual artifacts can implement those methods and allow more customization. I will fill in more details on the design proposal and use cases on that issue soon.
> 
> I'm not sure building an async version of BentoAPIServer solves that issue, but it is an interesting idea. BentoML currently has an async layer built with aiohttp that does request handling and micro-batching(https://docs.bentoml.org/en/latest/guides/micro_batching.html). I think this can be useful when the user does not need micro-batching. Although the user may lose the ability to run multiple model workers.

Thank you for letting me know about all this! I am starting to see that just using an asynchronous server does not fix the inherent issue, but I have a few other ideas for how to accomplish this. If you have already had some thoughts, I would love to hear them!
1. Make the service thread safe, so that the API server can be run in a background thread
2. Probably no code changes required, but a packed service can be restarted behind a load balancer if there are multiple instances of the server and switched out with zero downtime using a [Kubernetes rolling update](https://kubernetes.io/docs/tutorials/kubernetes-basics/update/update-intro/)

**Comment by parano** [comment]
hi, @akainth015  I think with tools like Kubernetes rolling update and other blue/green deployment tools available, it is already possible for DevOps to swap the BentoML API server container behind an endpoint and it is the recommended approach, see https://github.com/bentoml/BentoML/issues/754#issuecomment-640177122  

IMHO the hot loading model implementation within the python process can be a simple blocking call, as it is meant for easy development/testing rather than being used in mission-critical production environments. In which case, a comprehensive CI/CD workflow is always preferred.

From what I heard, the other issues mentioned in https://github.com/bentoml/BentoML/issues/758 especially reloading artifacts while developing BentoService is actually a lot more important for the project, because the user may need to re-train the model and re-run the packaging code every time they make some code change in their BentoService code. And it slows down the development workflow a lot.  I will put some notes from previous discussions on that issue and share it with you soon.

**Pull request closed without merging**